### PR TITLE
add continual learning to catboost and friends

### PR DESCRIPTION
**Pull request by robcaulk**
Add the ability for Catboost and friends to employ "continual learning." This feature allows users to tell FreqAI to continue training existing models from their most recently trained state. As opposed to the existing paradigm where a new model is trained each retrain.

Users can activate this functionality via:

`"continual_learning": true`

In their config files.

### CODE REVIEW

File: freqtrade/freqai/freqai_interface.py [hunk 1]
```diff
@@ -674,7 +675,7 @@ def train(self, unfiltered_dataframe: DataFrame, pair: str, dk: FreqaiDataKitche
         """
 
     @abstractmethod
-    def fit(self, data_dictionary: Dict[str, Any]) -> Any:
+    def fit(self, data_dictionary: Dict[str, Any], dk: FreqaiDataKitchen) -> Any:
```
Comment (xmatthias) [inline freqtrade/freqai/freqai_interface.py]: do we assume we'll have the need to extend this interface further?

if yes - i think we should consider recommending people add either `*args, **kwargs` - or just `**kwargs` to theyr implementation (as we do with strategy methods) - so we can extend the interface at will without breaking people's models over and over again.

Please note: 
the latter approach (`**kwargs` only) means we MUST call all additional arguments as keyword arguments (no longer as positional arguments).

Comment (robcaulk) [reply freqtrade/freqai/freqai_interface.py]: sounds good

File: freqtrade/freqai/prediction_models/LightGBMRegressorMultiTarget.py [hunk 2]
```diff
@@ -31,6 +32,9 @@ def fit(self, data_dictionary: Dict) -> Any:
         eval_set = (data_dictionary["test_features"], data_dictionary["test_labels"])
         sample_weight = data_dictionary["train_weights"]
 
+        if self.continual_learning:
+            logger.warning('Continual learning not supported for MultiTarget models')
```
Comment (xmatthias) [inline freqtrade/freqai/prediction_models/LightGBMRegressorMultiTarget.py]: wouldn't you kindof want to warn about this on startup?
I think with this it's going to be shown once per pair per training ...

Comment (robcaulk) [reply freqtrade/freqai/prediction_models/LightGBMRegressorMultiTarget.py]: I just went ahead and made our own FreqAI regressor to allow us to support continual learning and eval sets on multioutput :-)

File: freqtrade/freqai/prediction_models/CatboostRegressor.py [hunk 3]
```diff
@@ -38,16 +38,13 @@ def fit(self, data_dictionary: Dict) -> Any:
...
         model = CatBoostRegressor(
             allow_writing_files=False,
             **self.model_training_parameters,
         )
 
-        model.fit(X=train_data, eval_set=test_data)
-
-        # some evidence that catboost pools have memory leaks:
-        # https://github.com/catboost/catboost/issues/1835
-        del train_data, test_data
```
Comment (xmatthias) [inline freqtrade/freqai/prediction_models/CatboostRegressor.py:50]: we no longer fear from this memory leak?

File: freqtrade/freqai/prediction_models/LightGBMRegressorMultiTarget.py [hunk 4]
```diff
@@ -28,12 +29,36 @@ def fit(self, data_dictionary: Dict) -> Any:
...
+                eval_sets[i] = (  # type: ignore
+                    data_dictionary["test_features"],
+                    data_dictionary["test_labels"].iloc[:, i]
+                )
+
+        init_model = self.get_init_model(dk.pair)
+        if init_model:
+            init_models = init_model.estimators_
+        else:
+            init_models = [None] * y.shape[1]
```
Comment (xmatthias) [inline freqtrade/freqai/prediction_models/LightGBMRegressorMultiTarget.py:50]: None - or NaN ? (just asking ...)

**Review by xmatthias — APPROVED** [review]
just 2 questions (which are not THAT important, and can be tackled afterwards) 

LGTM :+1:

File: freqtrade/freqai/prediction_models/CatboostRegressor.py (hunk 3 above)
Comment (robcaulk) [reply freqtrade/freqai/prediction_models/CatboostRegressor.py:50]: That leak was solved with my earlier mem-leak solution :-)

File: freqtrade/freqai/prediction_models/LightGBMRegressorMultiTarget.py (hunk 4 above)
Comment (robcaulk) [reply freqtrade/freqai/prediction_models/LightGBMRegressorMultiTarget.py:50]: None !

**Pull request merged**